error_formatter.go

```diff
@@ -18,25 +18,50 @@ package jsonnet
 import (
 	"bytes"
 	"fmt"
+	"io"
 
-	"github.com/fatih/color"
 	"github.com/google/go-jsonnet/ast"
 	"github.com/google/go-jsonnet/parser"
 )
 
-type ErrorFormatter struct {
-	// MaxStackTraceSize  is the maximum length of stack trace before cropping
-	MaxStackTraceSize int
+type ErrorFormatter interface {
+	// Format static, runtime, and unexpected errors prior to printing them.
+	Format(err error) string
+
+	// Set the the maximum length of stack trace before cropping.
+	SetMaxStackTrackeSize(size int)
+
+	// Set the color formatter for the location color.
+	SetColorFormatter(color ColorFormatter)
+}
+
+type ColorFormatter func(w io.Writer, f string, a ...interface{}) (n int, err error)
+
+var _ ErrorFormatter = &termErrorFormatter{}
+
+type termErrorFormatter struct {
+	// maxStackTraceSize  is the maximum length of stack trace before cropping
+	maxStackTraceSize int
 
 	// Examples of current state of the art.
 	// http://elm-lang.org/blog/compiler-errors-for-humans
 	// https://clang.llvm.org/diagnostics.html
-	pretty   bool
-	colorful bool
-	SP       *ast.SourceProvider
+	color  ColorFormatter
+	pretty bool
+
+	// sp is currently never set, but is used to format locations.
```

Comment:
This may be very confusing. We could replace SourceProvider methods with ordinary functions. The original idea was to keep the source code in SourceProvider, but in the end the reference to the code is kept in each LocationRange anyway... But I think we can leave it as-is for now.